pennylane/dla/lie_closure.py

```diff
@@ -0,0 +1,282 @@
+
+    .. code-block::python3
+        ops = [
+            PauliSentence({
+                PauliWord({0:"X", 1:"X"}) : 1.,
+                PauliWord({0:"Y", 1:"Y"}) : 1.
+            }),
+            PauliSentence({
+                PauliWord({0:"X", 1:"X"}) : 1.,
+            }),
+            PauliSentence({
+                PauliWord({0:"Y", 1:"Y"}) : 1.,
+            }),
+        ]
+
+        vspace = VSpace(ops)
+
+    It automatically detects that the third operator is linearly dependent on the former two, so it does not add the third operator to the basis.
+
+    >>> vspace.basis
+    [1.0 * X(0) @ X(1)
+     + 1.0 * Y(0) @ Y(1),
+     1.0 * X(0) @ X(1)]
+
+    We can also retrospectively add operators.
+
+    >>> vspace.add(PauliWord({0:"X"}))
+    [1.0 * X(0) @ X(1)
+     + 1.0 * Y(0) @ Y(1),
+     1.0 * X(0) @ X(1),
+     1.0 * X(0)]
+
+    Again, checks of linear independence are always performed. So in the following example no operator is added.
+
+    >>> vspace.add(PauliWord({0:"Y", 1:"Y"}))
+    [1.0 * X(0) @ X(1)
+     + 1.0 * Y(0) @ Y(1),
+     1.0 * X(0) @ X(1),
+     1.0 * X(0)]
+    """
+
+    def __init__(self, generators):
+
+        # Get all Pauli words that are present in at least one Pauli sentence
+        all_pws = list(reduce(set.__or__, [set(ps.keys()) for ps in generators]))
+        num_pw = len(all_pws)
+        # Create a dictionary mapping from PauliWord to row index
+        self._pw_to_idx = {pw: i for i, pw in enumerate(all_pws)}
+
+        # Initialize VSpace properties trivially
+        self._basis = []
+        rank = 0
+
+        self._M = np.zeros((num_pw, rank), dtype=float)
+        self._rank = rank
+        self._num_pw = num_pw
+
+        # Add all generators that are linearly independent
+        self.add(generators)
+
+    @property
+    def basis(self):
+        """List of basis operators of VSpace"""
+        return self._basis
+
+    def __len__(self):
+        return len(self.basis)
+
+    def add(self, other):
+        """Adding a list of PauliSentences if they are linearly independent"""
+        if isinstance(other, (PauliWord, PauliSentence, Operator)):
+            other = [other]
+
+        other = [
+            qml.pauli.pauli_sentence(op) if not isinstance(op, PauliSentence) else op
+            for op in other
+        ]
+
+        for ps in other:
+            # TODO: Potential speed-up by computing the maximal linear independent set for all current basis vectors + other, essentially algorithm1 in https://arxiv.org/abs/1012.5256
+            self._M, self._pw_to_idx, self._rank, self._num_pw, is_independent = (
+                self._add_if_independent(self._M, ps, self._pw_to_idx, self._rank, self._num_pw)
+            )
+            if is_independent:
+                self._basis.append(ps)
+        return self._basis
+
+    @staticmethod
+    def _add_if_independent(M, pauli_sentence, pw_to_idx, rank, num_pw):
```

Comment:
I have the thought of trying to avoid in-place manipulations whenever possible but cant really pin it down to something concrete (like a rule or PEP) - except for jax-compatibility maybe?

Would actually appreciate some input here, also in general how to best handle private and public attributes of the class

Reply:
You're using this static method in `PauliVSpace.add` to update class attributes, so you're performing in-place mutation. However, with the changes you've added recently, I think it makes sense to leave this as a static method.

Regarding private/public attributes, it just comes down to what you want the user to be able to see and how you want to express it. You might have private attributes that are helpful for devs but complete garbage for users, but you can use it to create meaningful user facing properties/functions. Sorry if I'm explaining stuff you already know, but I think your implementation is sound. You've hidden away unnecessary details well enough in private methods like the above.